<!DOCTYPE html>
<html>
<title>ChromaClade</title>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="{{url_for('static', filename='w3.css')}}">
<link rel="stylesheet" href="{{url_for('static', filename='montserrat.css')}}">
<link rel="stylesheet" href="{{url_for('static', filename='font-awesome.min.css')}}">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
<link rel="stylesheet" href="{{url_for('static', filename='custom.css')}}" >
<!--TODO can't get local download of fa css to work yet - may need to download more of their files? if it matters-->
<style>
    body, h1,h2,h3,h4,h5,h6 {font-family: "Montserrat", sans-serif}
    .w3-row-padding img {margin-bottom: 12px}
    /* Set the width of the sidebar to 120px */
    .w3-sidebar {width: 120px;background: #222;}
    /* Add a left margin to the "page content" that matches the width of the sidebar (120px) */
    #main {margin-left: 120px}
    /* Remove margins from "page content" on small screens */
    @media only screen and (max-width: 600px) {
        #main {margin-left: 0}
    }
</style>

<body class="w3-white">

<!-- Icon Bar (Sidebar - hidden on small screens) -->
<nav class="w3-sidebar w3-bar-block w3-small w3-hide-small w3-center w3-white" style="border: 1px solid silver;">
  <!-- Avatar image in top left corner -->
  <img src="{{url_for('static', filename='tree.png')}}" style="width:100%" class="w3-white">
  <a href="#" class="w3-bar-item w3-button w3-padding-large w3-hover-black">
    <i class="fa fa-home w3-xxlarge"></i>
    <p>ChromaClade</p>
  </a>
<a href="#about" class="w3-bar-item w3-button w3-padding-large w3-hover-black">
    <i class="fa fa-user w3-xxlarge"></i>
    <p>About</p>
  </a>
  <a href="#how-to" class="w3-bar-item w3-button w3-padding-large w3-hover-black">
    <i class="fa fa-book w3-xxlarge"></i>
    <p>How to</p>
  <a href="#examples" class="w3-bar-item w3-button w3-padding-large w3-hover-black">
    <i class="fa fa-eye w3-xxlarge"></i>
    <p>Examples</p>
  </a>
  
  </nav>

<!-- Navbar on small screens (Hidden on medium and large screens) -->
<div class="w3-top w3-hide-large w3-hide-medium" id="myNavbar">
  <div class="w3-bar w3-white w3-opacity w3-hover-opacity-off w3-center w3-small">
    <a href="#" class="w3-bar-item w3-button" style="width:25% !important">ChromaClade</a>
    <a href="#about" class="w3-bar-item w3-button" style="width:25% !important">About</a>
    <a href="#how-to" class="w3-bar-item w3-button" style="width:25% !important">How to</a>
    <a href="#examples" class="w3-bar-item w3-button" style="width:25% !important">Examples</a>
  </div>
</div>

<!-- Page Content -->
  <!-- Header/Home -->
<header class="w3-container w3-padding-16 w3-center w3-white" id="home">
    <h1 class="w3-jumbo"> ChromaClade </h1>
    <span>Combined visualisation of phylogenetic and sequence data</span>
    <!--    <img src="{{url_for('static', filename='tree.png')}}" alt="coloured tree" class="w3-image w3-opacity-min" width="100" height="100">-->
</header>

<main>
{% block main %}{% endblock %}
</main>

<p class="w3-center self-promotion">
    ChromaClade is developed and maintained by <a href="https://uk.linkedin.com/in/chrismonit" target="_blank">Christopher Monit</a>. It is the result of my past postdoctoral research work in the <a href="https://www.ucl.ac.uk/towers-lab/" target="_blank">Greg Towers laboratory</a>.
    <!-- TODO and funded by... -->
</p>

<i class="fa fa-book"></i>


<div id="about" class="about">
    <h1>About</h1>
    <p>
    ChromaClade is a bioinformatics program that produces visualisations combining sequence and phylogenetic data. It annotates taxon names in a phylogenetic tree with amino acids found at individual sites in a corresponding sequence alignment. Residue-specific colour codes enable easy inspection by eye and the resulting trees can be viewed in any of several popular tree-viewing programs.
    </p>

    <img src="{{url_for('static', filename='explanation.png')}}" alt="ChromaClade summary" class="w3-image w3-opacity-min">




    <!-- TODO how to cite, link to towers lab and contact -->
</div>
       
<div id="how-to" class="how-to">
    <h1>Using ChromaClade</h1>
    <!-- TODO add instructions about making alignments and trees -->

    <h3>Input</h3>
    <p>
Use the buttons to select the files containing your tree and your corresponding alignment. 
    </p>

    <p>
    Use the dropdown to select the <b>file format</b> for your input files: either <a href="https://en.wikipedia.org/wiki/Newick_format" target="_blank">Newick</a> or <a href="https://en.wikipedia.org/wiki/Nexus_file" target="_blank">Nexus</a> format for trees and <a href="https://en.wikipedia.org/wiki/FASTA_format" target="_blank">Fasta</a> or Nexus format for alignment. Tools are available for converting between tree formats (e.g. <a href="http://phylogeny.lirmm.fr/phylo_cgi/data_converter.cgi" target="_blank">phylogeny.fr</a>) and alignment formats (e.g. <a href="https://www.ebi.ac.uk/Tools/sfc/emboss_seqret/" target="_blank">EBI tools</a>).
    </p>

    <p>
The alignment would ordinarily be amino acid sequences (e.g. translated from the nucleotide sequences used to first estimate the tree) but the method will work perfectly well with nucleotide states too. 
    </p>

    <h3>Output</h3>

    <p>
The output is a set of trees saved to a single file, where taxon names have been annotated according to the amino acid found in that taxon's sequence. E.g. if methionine is found at position 1 in the human sequence, then <code>human</code> becomes <code>human__site_1__M</code>. In addition, the taxon name will be assigned a colour specific to methionine.
    </p>
    
    <p>
Use the dropdown to select the <b>output file format</b> for the output file, either a Nexus format file specifically compatible with the <a href="http://tree.bio.ed.ac.uk/software/figtree/" target="_blank">FigTree</a> tree viewer, or the PhyloXML format compatible with other viewers such as <a href="https://sites.google.com/site/cmzmasek/home/software/archaeopteryx" target="_blank">Archaeopteryx</a>.
    </p>

    <p>
Branches can also be coloured by amino acids observed in descendent taxa, by clicking the <b>Colour branches</b> checkbox. Annotated trees can be made for a subset of sites by clicking the <b>Choose sites</b> checkbox and entering the site ranges to include, just as you would specify pages of a document to print.
    </p>

    <!-- TODO Note about how long it stays on server? -->

</div>


<div id="examples" class="examples">
    <h1>Examples</h1>  <!-- TODO use copy from paper, paraphrased -->
    These are sequences of the PB2 protein found in influenza A viruses infecting humans and birds. Below is a colour-annotated tree produced by ChromaClade and visualised in <a href="http://tree.bio.ed.ac.uk/software/figtree/" target="_blank">FigTree</a>, showing residues observed at site 271, namely threonine (blue), alanine (red) and isoleucine (purple).

    <p align="center">
        <img src="{{url_for('static', filename='pb2_271.jpg')}}" alt="PB2 site 271" width="300"/>
    </p>

</div>



<!--TODO: make a footer-->

</body>
</html>
